ide/vscoq/vscoqtop.ml

```diff
@@ -11,285 +11,33 @@
-
-let textDocumentDidOpen params : unit Lwt.t =
-  let open Yojson.Basic.Util in
-  let open Lwt.Infix in
-  let textDocument = params |> member "textDocument" in
-  let uri = textDocument |> member "uri" |> to_string in
-  let text = textDocument |> member "text" |> to_string in
-  let document = create_document (get_init_state ()) text in
-  Hashtbl.add documents uri document;
-  send_highlights uri document <&>
-  publish_diagnostics uri document
-
-
-let textDocumentDidChange params : unit Lwt.t =
-  let open Yojson.Basic.Util in
-  let open Lwt.Infix in
-  let textDocument = params |> member "textDocument" in
-  let uri = textDocument |> member "uri" |> to_string in
-  let contentChanges = params |> member "contentChanges" |> to_list in
-  let read_edit edit =
-    let text = edit |> member "text" |> to_string in
-    let range = edit |> member "range" in
-    let range_start = range |> member "start" |> parse_loc in
-    let range_end = range |> member "end" |> parse_loc in
-    DocumentManager.({ range_start; range_end }, text)
-  in
-  let textEdits = List.map read_edit contentChanges in
-  let document = Hashtbl.find documents uri in
-  let new_doc = DocumentManager.apply_text_edits document textEdits in
-  Hashtbl.replace documents uri new_doc;
-  send_highlights uri new_doc <&>
-  publish_diagnostics uri new_doc
-
-let textDocumentDidSave params : unit Lwt.t =
-  let open Yojson.Basic.Util in
+let loop run_mode ~opts:_ state =
+  LspManager.init ();
   let open Lwt.Infix in
-  let textDocument = params |> member "textDocument" in
-  let uri = textDocument |> member "uri" |> to_string in
-  let document = Hashtbl.find documents uri in
-  let new_doc = DocumentManager.validate_document document in
-  Hashtbl.replace documents uri new_doc;
-  send_highlights uri new_doc <&>
-  publish_diagnostics uri new_doc
-
-let mk_goal sigma g =
-  let env = Goal.V82.env sigma g in
-  let min_env = Environ.reset_context env in
-  let id = Goal.uid g in
-  let ccl =
-    pr_letype_env ~goal_concl_style:true env sigma (Goal.V82.concl sigma g)
-  in
-  let mk_hyp d (env,l) =
-    let d' = CompactedDecl.to_named_context d in
-    let env' = List.fold_right Environ.push_named d' env in
-    let ids, body, typ = match d with
-    | CompactedDecl.LocalAssum (ids, typ) ->
-       ids, None, typ
-    | CompactedDecl.LocalDef (ids,c,typ) ->
-      ids, Some c, typ
-    in
-  let ids = List.map (fun id -> `String (Names.Id.to_string id.Context.binder_name)) ids in
-  let typ = pr_ltype_env env sigma typ in
-    let hyp = `Assoc ([
-      "identifiers", `List ids;
-      "type", `String (Pp.string_of_ppcmds typ);
-      "diff", `String "None";
-    ] @ Option.cata (fun body -> ["body", `String (Pp.string_of_ppcmds @@ pr_lconstr_env ~inctx:true env sigma body)]) [] body) in
-    (env', hyp :: l)
+  let rec loop (events : 'a LspManager.events) =
+    let open Yojson.Basic.Util in
+    log @@ "[T] looking for next step";
+    Lwt_io.flush_all () >>= fun () ->
```

Comment:
This is the final "main loop" which does not look too ugly to me

Reply:
The alternative is to hide sub components' events inside their super component.
In this way we don't need polymorphic variants, but you would call lspmanager.handle_event always, and in some cases that handle event would dispatch to other handle_event (like the one of the delegation manager or the one of the tactic manager if we implement par:).

Reply:
I've implemented the sealing, the main loop is now agnostic on the component/type of events